### PR TITLE
Add a webhook to process pipeline event

### LINKED ISSUE #325: PipelineRun will still be repeated here
Please see the repeated PipelineRuns:

image

**How to reproduce?**

1. Create a Pipeline which Jenkinsfile is as below:

```groovy
pipeline {
    agent any
    stages {
        stage('Non-Parallel Stage') {
            steps {
                echo 'This stage will be executed first.'
            }
        }
        stage('Parallel Stage') {
            when {
                branch 'master'
            }
            failFast true
            parallel {
                stage('Branch A') {
                    agent {
                        label "for-branch-a"
                    }
                    steps {
                        echo "On Branch A"
                    }
                }
                stage('Branch B') {
                    agent {
                        label "for-branch-b"
                    }
                    steps {
                        echo "On Branch B"
                    }
                }
            }
        }
    }
}
```

2. Click `Run` button to create PipelineRun

> Please note that this scenario does not happen every time. But more will happen for the first time.

/kind bug
/priority high
/area devops
/assign
/milestone v3.2

### PR DESCRIPTION
**Pull request by JohnNiang**
### What type of PR is this?
/kind feature

### What this PR does / why we need it:

1. Add a webhook to process pipeline event
    1. Create PipelineRun when event type is `run.initialize`
2. Remove the request of synchronizing PipelineRun and disable PipelineRun synchronizer by default

Please refer to 
- https://github.com/kubesphere/ks-devops/discussions/433
- https://github.com/kubesphere/ks-devops/issues/65
### Which issue(s) this PR fixes:

Fixes #325

### Special notes for reviewers:

Docker images for test
```
# DevOps Jenkins
kubespheredev/ks-jenkins:master
# DevOps APIServer
johnniang/devops-apiserver:dev-v3.2.1-54886e1
# DevOps Controller
johnniang/devops-controller:dev-v3.2.1-54886e1
```

Steps to test:

- Replace docker images above for Deployment devops-apiserver and devops-controller
- Create a simple Pipeline in KubeSphere console
- Trigger the corresponding Pipeline in Jenkins
- See the PipelineRuns in KubeSphere console

```go
 W0118 02:18:45.926535       1 client_config.go:552] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.

 W0118 02:18:45.928388       1 options.go:141] ks-apiserver starts without redis provided, it will use in memory cache. This may cause inconsistencies when running ks-apiserver with multiple replicas.

 I0118 02:18:45.971727       1 sonarqube.go:39] Sonarqube integration is disabled

 I0118 02:18:45.973780       1 sonarqube.go:39] Sonarqube integration is disabled

 [restful] 2022/01/18 02:18:46 log.go:33: [restful/swagger] listing is available at /apidocs.json

 [restful] 2022/01/18 02:18:46 log.go:33: [restful/swagger] /apidocs/ is mapped to folder bin/swagger-ui

 I0118 02:18:46.004237       1 apiserver.go:220] Start cache objects

 I0118 02:18:46.226483       1 apiserver.go:275] Finished caching objects

 I0118 02:18:46.226576       1 apiserver.go:184] Start listening on :9090

 I0118 02:25:25.826024       1 workflow_handler.go:116] Created a PipelineRun: my-devops-projectdq8jj/simple-pipeline-qst5s

 I0118 02:26:02.346993       1 workflow_handler.go:116] Created a PipelineRun: my-devops-projectdq8jj/simple-pipeline-vkvcz

 I0118 02:26:07.726954       1 workflow_handler.go:116] Created a PipelineRun: my-devops-projectdq8jj/simple-pipeline-fxpcv
```

TODOs:
- [x] https://github.com/kubesphere/ks-jenkins/pull/77
- [x] https://github.com/kubesphere-sigs/ks-devops-helm-chart/pull/53

### Does this PR introduce a user-facing change??

```release-note
Add a webhook to process pipeline event
```

/label tide/merge-method-squash
/cc @kubesphere/sig-devops

### CODE REVIEW

**Comment by JohnNiang** [comment]
/hold
I hold this PR util [pipeline-event-plugin](https://github.com/johnniang/pipeline-event-plugin) is ready and added into [ks-jenkins](https://github.com/kubesphere/ks-jenkins).

**Comment by JohnNiang** [comment]
Hi @LinuxSuRen , I verified a problem by configuring an unknown configuration item (invalidEventDispatcher) in CasC. Finally, Jenkins could not load the CasC configuration.

```java
Jan 16, 2022 2:13:45 PM WARNING hudson.init.impl.InstallUncaughtExceptionHandler handleException
Caught unhandled exception with ID 5e23318f-7e18-4125-af36-28d0154fcdbd
io.jenkins.plugins.casc.ConfiguratorException: Invalid configuration elements for type class jenkins.model.GlobalConfigurationCategory$Unclassified : invalidEventDispatcher.
Available attributes : administrativeMonitorsConfiguration, appInsightsGlobalConfig, artifactManager, bitbucketEndpointConfiguration, buildDiscarders, buildStepOperation, builtInNode, cVSSCM, casCGlobalConfig, defaultFolderConfiguration, defaultView, envVarsFilter, eventDispatcher, fingerprints, gitHubConfiguration, gitHubPluginConfig, gitLabServers, gitSCM, globalConfigFiles, globalConfigurationImpl, globalDefaultFlowDurabilityLevel, globalLibraries, globalSettings, jiraGlobalConfiguration, junitTestResultStorage, kubesphereTokenAuthGlobalConfiguration, location, lockableResourcesManager, mailer, mavenModuleSet, metricsAccessKey, myView, nodeProperties, pipeline-model-docker, plugin, pollSCM, projectNamingStrategy, prometheusConfiguration, quietPeriod, resourceRoot, scmRetryCount, shell, singleConditionalBuilder, sonarGlobalConfiguration, subversionSCM, usageStatistics, viewsTabBar, whitelist
	at io.jenkins.plugins.casc.BaseConfigurator.handleUnknown(BaseConfigurator.java:375)
	at io.jenkins.plugins.casc.BaseConfigurator.configure(BaseConfigurator.java:364)
	at io.jenkins.plugins.casc.BaseConfigurator.check(BaseConfigurator.java:286)
	at io.jenkins.plugins.casc.ConfigurationAsCode.lambda$checkWith$8(ConfigurationAsCode.java:776)
	at io.jenkins.plugins.casc.ConfigurationAsCode.invokeWith(ConfigurationAsCode.java:712)
Caused: io.jenkins.plugins.casc.ConfiguratorException: unclassified: error configuring 'unclassified' with class io.jenkins.plugins.casc.impl.configurators.GlobalConfigurationCategoryConfigurator configurator
	at io.jenkins.plugins.casc.ConfigurationAsCode.invokeWith(ConfigurationAsCode.java:718)
	at io.jenkins.plugins.casc.ConfigurationAsCode.checkWith(ConfigurationAsCode.java:776)
	at io.jenkins.plugins.casc.ConfigurationAsCode.configureWith(ConfigurationAsCode.java:761)
	at io.jenkins.plugins.casc.ConfigurationAsCode.configureWith(ConfigurationAsCode.java:637)
	at io.jenkins.plugins.casc.ConfigurationAsCode.configure(ConfigurationAsCode.java:306)
```

File: pkg/api/devops/v1alpha3/pipelinerun_types.go [hunk 1]
```diff
@@ -167,6 +168,35 @@ func (pr *PipelineRun) GetPipelineRunID() (pipelineRunID string, exist bool) {
 	return
 }
 
+// GetPipelineRunIdentifier returns an identifier string which can identify a PipelineRun,
+// even if the PipelineRun belongs to multi-branch or non multi-branch Pipeline.
+// Format of the string: [scm.ref-name]-runID
+// TODO Add unit test aginst this method
```
Comment (LinuxSuRen) [inline pkg/api/devops/v1alpha3/pipelinerun_types.go:174]: typo `aginst`

File: pkg/kapis/devops/v1alpha3/webhook/workflow_handler.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+)
+
+type pipelineRunIdentifier struct {
+	namespaceName string
+	pipelineName  string
+	scmRefName    string
+	buildNumber   string
+}
+
+func (identifier *pipelineRunIdentifier) toIdentifier() string {
```
Comment (LinuxSuRen) [inline pkg/kapis/devops/v1alpha3/webhook/workflow_handler.go:38]: How about renaming this to `String()`.

For example: `func (identifier *pipelineRunIdentifier) String() string {`

the struct name already is an `xxIdentifier`, it makes me feel confused about `toIdentifier`. It already is `identifier`. do we need turn to `identifier` again?

File: pkg/kapis/devops/v1alpha3/webhook/register.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+	"kubesphere.io/devops/pkg/api"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// RegisterWebhooks registers all webhooks into web service.
+func RegisterWebhooks(genericClient client.Client, ws *restful.WebService) {
+	webhookHandler := NewHandler(genericClient)
+	ws.Route(ws.POST("/webhooks/jenkins").
+		To(webhookHandler.ReceiveEventsFromJenkins).
+		Doc("Webhook for receiving events from Jenkins").
```
[the remark this answers is not in the document]
Comment (JohnNiang) [reply pkg/kapis/devops/v1alpha3/webhook/register.go:32]: We have no need to return anything, but only HTTP status 200 to tell Jenkins we have received and handled event correctly. So I set the response data structure to nil.

File: pkg/event/models/common/workflowrun_handler.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	}
+	if event.TypeEquals(RunFinalized) && funcs.HandleFinalized != nil {
+		return funcs.HandleFinalized(workflowRunData)
+	}
+	if event.TypeEquals(RunCompleted) && funcs.HandleCompleted != nil {
+		return funcs.HandleCompleted(workflowRunData)
+	}
+	if event.TypeEquals(RunDeleted) && funcs.HandleDeleted != nil {
+		return funcs.HandleDeleted(workflowRunData)
+	}
```
[the remark this answers is not in the document]
Comment (JohnNiang) [reply pkg/event/models/common/workflowrun_handler.go:53]: Of course. I have refactored those ugly if checks.

File: pkg/kapis/devops/v1alpha3/webhook/register.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (LinuxSuRen) [reply pkg/kapis/devops/v1alpha3/webhook/register.go:32]: > but only HTTP status 200 to tell Jenkins

How about documenting it with the restFul style?

File: pkg/event/workflowrun/workflowrun_handler.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+		common.RunStarted:    handlers.HandleStarted,
+		common.RunFinalized:  handlers.HandleFinalized,
+		common.RunCompleted:  handlers.HandleCompleted,
+		common.RunDeleted:    handlers.HandleDeleted,
+	}
+	handler, exist := handlerMap[eventType]
+	if !exist || handler == nil {
+		return discardHandler
+	}
+	return handler
```
Comment (LinuxSuRen) [inline pkg/event/workflowrun/workflowrun_handler.go:57]: Looks better.

File: pkg/kapis/devops/v1alpha3/webhook/workflowrun_handler.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"context"
+	"fmt"
+	workflowrun2 "kubesphere.io/devops/pkg/event/workflowrun"
```
Comment (LinuxSuRen) [inline pkg/kapis/devops/v1alpha3/webhook/workflowrun_handler.go:22]: Missing here? I prefer to have a better name instead of `xxxx2`.

**Review by LinuxSuRen — CHANGES_REQUESTED** [review]

File: pkg/kapis/devops/v1alpha3/webhook/workflowrun_handler.go (hunk 6 above)
Comment (JohnNiang) [reply pkg/kapis/devops/v1alpha3/webhook/workflowrun_handler.go]: Thanks for your reminder! I changed that weird import alias in workflowrun_handler_test.go but forgot workflowrun_handler.go. For the latest fix please see https://github.com/kubesphere/ks-devops/pull/442/commits/88daab70905b9f56c6b39ba9d5526b80dfa0e7ac

**Comment by LinuxSuRen** [comment]
Below is the image of this PR:
```
kubespheredev/devops-apiserver:dev-v3.2.1-88daab7
```

**Comment by mzmuer** [comment]
I have tested it manually. It works normal. 

OS: `darwin/arm64`
images:
```
kubespheredev/devops-apiserver:dev-v3.2.1-88daab7
kubespheredev/ks-jenkins:master
johnniang/devops-controller:dev-v3.2.1-54886e1
```

**Comment by JohnNiang** [comment]
> I have tested it manually. It works normal.
> 
> OS: `darwin/arm64` images:
> 
> ```
> kubespheredev/devops-apiserver:dev-v3.2.1-88daab7
> kubespheredev/ks-jenkins:master
> johnniang/devops-controller:dev-v3.2.1-54886e1
> ```

Thanks for your patient testing. I will fix the conflicts later.

**Review by LinuxSuRen — APPROVED** [review]

**Pull request merged**